src/colorizer/VectorField.ts

```diff
@@ -0,0 +1,313 @@
+    if (range && this.config.visible) {
+      this.line.geometry.setDrawRange(range[0], range[1] - range[0]);
+    } else {
+      this.line.geometry.setDrawRange(0, 0);
+    }
+  }
+
+  /**
+   * Scales from frame pixel coordinates to relative frame coordinates.
+   *
+   * Frame pixel coordinates are in a range from `[0, frameResolution.x]` and
+   * `[0, frameResolution.y]`, where `(0, 0)` is the top left corner.
+   * Relative frame coordinates are in a range from `[-0.5, 0.5]`, where
+   * `(0, 0)` is the center of the frame and `(-0.5, 0.5)` is the top left corner.
+   */
+  private framePixelsToRelativeFrameCoords(vector: ArrayVector3): ArrayVector3 {
+    if (!this.dataset) {
+      return [0, 0, 0];
+    }
+    return [
+      (vector[0] / this.dataset.frameResolution.x) * 2.0 - 1.0,
+      -(vector[1] / this.dataset.frameResolution.y) * 2.0 + 1.0,
+      vector[2],
+    ];
+  }
+
+  /**
+   * Scales from relative frame coordinates to screen space pixel coordinates.
+   * Relative frame coordinates are in a range from `[-0.5, 0.5]`, and
+   * screen space pixel coordinates match the onscreen canvas pixels.
+   */
+  private relativeFrameCoordsToScreenSpacePx(vector: ArrayVector3): ArrayVector3 {
+    return [
+      vector[0] * this.frameToCanvasCoordinates.x * this.canvasResolution.x,
+      vector[1] * this.frameToCanvasCoordinates.y * this.canvasResolution.y,
+      vector[2],
+    ];
+  }
+
+  private getArrowHeadVertices(normVectorStart: ArrayVector3, normVectorEnd: ArrayVector3): Float32Array {
+    //    o <- normVectorEnd
+    //   /|\
+    //  / | \  <- arrow head maintains onscreen pixel length with zoom
+    //    |
+    //    |
+    //    o <- normVectorStart
+
+    // Draw the arrow heads. These are two line segments for each, so there's a total of four additional vertices we need to add.
+    const vertices = new Float32Array((VERTICES_PER_VECTOR_LINE - 2) * VERTEX_LENGTH);
+
+    // TODO: Handle scaling and arrow head length in a vertex shader.
+    // There's extra work being done here to keep the arrow head length constant with zoom, but it also
+    // means that currently this is recalculated EVERY time the zoom changes.
+    const screenSpaceStartPx = this.relativeFrameCoordsToScreenSpacePx(normVectorStart);
+    const screenSpaceEndPx = this.relativeFrameCoordsToScreenSpacePx(normVectorEnd);
+    const screenSpaceDeltaPx = [
+      screenSpaceEndPx[0] - screenSpaceStartPx[0],
+      screenSpaceEndPx[1] - screenSpaceStartPx[1],
+    ];
+    const vectorLengthPx = Math.sqrt(
+      screenSpaceDeltaPx[0] * screenSpaceDeltaPx[0] + screenSpaceDeltaPx[1] * screenSpaceDeltaPx[1]
+    );
+
+    const vectorAngle = Math.atan2(screenSpaceDeltaPx[1], screenSpaceDeltaPx[0]) + Math.PI;
+    const arrowAngle1 = vectorAngle + arrowStyle.headAngleRads;
+    const arrowAngle2 = vectorAngle - arrowStyle.headAngleRads;
+    const arrowHead1: ArrayVector3 = [Math.cos(arrowAngle1), Math.sin(arrowAngle1), 0];
```

Comment:
are these the ending points of the arrow line segment? `arrowHead1` and `arrowHead2` could use better names. I would also consider using left and right instead of 1 and 2